firedrake/preconditioners/fdm.py

```diff
@@ -852,15 +874,48 @@ def condense(self, result=None):
 
 
 class SchurComplementBlockCholesky(SchurComplementKernel):
```

Comment:
I feel like all of these classes should have proper docstrings explaining what they do. I at least do not have the mathematical background to know what they do just from the names.